src/ast/irbuilderbpf.cpp

```diff
@@ -543,6 +543,90 @@ void IRBuilderBPF::CreateForEachMapElem(Value *ctx,
   CreateHelperErrorCond(ctx, call, libbpf::BPF_FUNC_for_each_map_elem, loc);
 }
 
+void IRBuilderBPF::CreateCheckSetRecursion(const location &loc)
+{
+  const std::string map_ident = to_string(MapType::RecursionPrevention);
+
+  AllocaInst *key = CreateAllocaBPF(getInt32Ty(), "lookup_key");
+  CreateStore(getInt32(0), key);
+
+  CallInst *call = createMapLookup(map_ident, key);
+
+  Function *parent = GetInsertBlock()->getParent();
+  BasicBlock *lookup_success_block = BasicBlock::Create(module_.getContext(),
+                                                        "lookup_success",
+                                                        parent);
+  BasicBlock *lookup_failure_block = BasicBlock::Create(module_.getContext(),
+                                                        "lookup_failure",
+                                                        parent);
+  BasicBlock *merge_block = BasicBlock::Create(module_.getContext(),
+                                               "merge",
+                                               parent);
+
+  // Make the verifier happy with a null check even though the value should
+  // never be null for key 0.
+  Value *condition = CreateICmpNE(
+      CreateIntCast(call, GET_PTR_TY(), true),
+      ConstantExpr::getCast(Instruction::IntToPtr, getInt64(0), GET_PTR_TY()),
+      "map_lookup_cond");
+  CreateCondBr(condition, lookup_success_block, lookup_failure_block);
+
+  SetInsertPoint(lookup_success_block);
+
+  CreateLifetimeEnd(key);
+
+  // createMapLookup  returns an u8*
+  auto *cast = CreatePointerCast(call, getInt64Ty(), "cast");
+
+  Value *prev_value = CREATE_ATOMIC_RMW(AtomicRMWInst::BinOp::Xchg,
+                                        cast,
+                                        getInt64(1),
+                                        8,
+                                        AtomicOrdering::SequentiallyConsistent);
+
+  Function *set_parent = GetInsertBlock()->getParent();
+  BasicBlock *value_is_set_block = BasicBlock::Create(module_.getContext(),
+                                                      "value_is_set",
+                                                      set_parent);
+  Value *set_condition = CreateICmpEQ(prev_value,
+                                      getInt64(1),
+                                      "value_set_condition");
+  CreateCondBr(set_condition, value_is_set_block, merge_block);
+
+  SetInsertPoint(value_is_set_block);
+  /*
+   * The counter is set, we need to exit early from the probe.
+   * This should only happen for the functions that can lead
+   * to a recursive crash e.g. "queued_spin_lock_slowpath"
+   */
+  CreateRet(getInt64(0));
```

Comment:
But this is a case of recursion right? Does that really qualify as a dropped event?

Reply:
I'm also not keen on adding more logic in an escape path where we're trying to prevent doing work. Maybe I'm being overly paranoid.

Reply:
At a high level (I and many others) believe that it should be a functional requirement that we don't ever lose an event. We may not always be able to execute code, but we need to at least report it.

Particularly in weird spots like these, it's critical we report back to user a corner case happened. Corner cases are typically why people use tracing over profiling.

Practically speaking, a percpu array access is rewritten to a direct memory access (no helpers involved). And we can do an atomic increment on the destination address. So it is a failure free codepath. If it is not a failure free codepath, the kernel must change to support this.

Reply:
Discussed offline. I'm good to add this accounting for lost events but I think it might need to be a separate PR as I'll need to make a more general version of `handle_ringbuf_loss`.

Reply:
Ok, making the ring buff loss counter generic was pretty easy. @danobi Let me know if you want me to stack the commit on top of this one in this PR. Example ouput:
```
[jordalgo@devbig1151.frc2 ~/local/bpftrace/build (kfunc-counter-3144)]$ sudo ./src/bpftrace ~/abba-crash-repro.bt
Attaching 2 probes...
Lost 13 events
Lost 28 events
Lost 7 events
Lost 8 events
Lost 322 events
Lost 14 events
Lost 1 events
Lost 31 events
...
```

Reply:
Follow up is fine!